pkg/cloud/services/provider/provider.go

```diff
@@ -109,9 +109,22 @@ func NewClient(cloud clientconfig.Cloud, caCert []byte) (*gophercloud.ProviderCl
 	}
 	err = openstack.Authenticate(provider, *opts)
 	if err != nil {
-		return nil, nil, fmt.Errorf("providerClient authentication err: %v", err)
+		return nil, nil, "", fmt.Errorf("providerClient authentication err: %v", err)
 	}
-	return provider, clientOpts, nil
+
+	// Set ProjectID for consumers of the ProviderClient
+	// This is only needed when application credentials were used, because they
+	// do not specify a `project_id` in the clientconfig.Cloud.
+	var projectID string
+	if tokenResult, ok := provider.GetAuthResult().(tokens.CreateResult); ok {
+		project, err := tokenResult.ExtractProject()
+		if err != nil {
+			return nil, nil, "", fmt.Errorf("unable to get project from createResult: %v", err)
+		}
+		projectID = project.ID
+	}
```

Comment:
I think we want to return an error here if we failed to get a project from the authentication response. This corresponds to the error we would previously have raised if we didn't get a project from clientopts.

Because we have always required project to be specified in clientopts, we have always required project-scoped authentication, so IIUC project should always be available.

https://docs.openstack.org/api-ref/identity/v3/?expanded=#authentication-and-token-management

Reply:
That makes sense. I will add an error return on the next push.